File: service-beacons-node/src/clients/version1/BeaconsCommandableGrpcClientV1.ts
import { ConfigParams, DataPage, FilterParams, PagingParams } from 'pip-services3-commons-node';
import { CommandableGrpcClient } from 'pip-services3-grpc-node';
import { BeaconV1 } from '../..';

import { IBeaconsClientV1 } from './IBeaconsClientV1';

export class BeaconsCommandableGrpcClientV1 extends CommandableGrpcClient implements IBeaconsClientV1 {
    private _blockCount: number = 1;
    private _ids: string[] = null;

    public constructor() {
        super('beacons_v1');
    }

    public configure(config: ConfigParams): void {
        super.configure(config);

        this._blockCount = config.getAsIntegerWithDefault("options.block_count", this._blockCount);
    }

    public getBeacons(correlationId: string, filter: FilterParams, paging: PagingParams,
        callback: (err: any, page: DataPage<BeaconV1>) => void): void {
        this.callCommand(
            'get_beacons',
            correlationId,
            {
                filter: filter,
                paging: paging
            },
            callback
        );
    }

    public getBeaconById(correlationId: string, beaconId: string,
        callback: (err: any, beacon: BeaconV1) => void): void {

        this.callCommand(
            'get_beacon_by_id',
            correlationId,
            {
                beacon_id: beaconId
            },
            callback
        );
    }

    public getBeaconByUdi(correlationId: string, udi: string,
        callback: (err: any, beacon: BeaconV1) => void): void {
        this.callCommand(
            'get_beacon_by_udi',
            correlationId,
            {
                udi: udi
            },
            callback
        );
    }

    public calculatePosition(correlationId: string, siteId: string, udis: string[],
        callback: (err: any, position: any) => void): void {
        this.callCommand(
            'calculate_position',
            correlationId,
            {
                site_id: siteId,
                udis: udis
            },
            callback
        );
    }

    public createBeacon(correlationId: string, beacon: BeaconV1,
        callback: (err: any, beacon: BeaconV1) => void): void {
        this.callCommand(
            'create_beacon',
            correlationId,
            {
                beacon: beacon
            },
            callback
        );
    }

    public updateBeacon(correlationId: string, beacon: BeaconV1,
        callback: (err: any, beacon: BeaconV1) => void): void {
        this.callCommand(
            'update_beacon',
            correlationId,
            {
                beacon: beacon
            },
            callback
        );
    }

    public deleteBeaconById(correlationId: string, beaconId: string,
        callback: (err: any, beacon: BeaconV1) => void): void {
        this.callCommand(
            'delete_beacon_by_id',
            correlationId,
            {
                beacon_id: beaconId
            },
            callback
        );
    }

}

File: service-beacons-dotnet/src/Client/Clients/Version1/BeaconsCommandableGrpcClientV1.cs
using System.Threading.Tasks;
using PipServices3.Grpc.Clients;
using Nov.MaxSamples.Beacons.Data.Version1;
using PipServices3.Commons.Data;

namespace Nov.MaxSamples.Beacons.Clients.Version1
{
	public class BeaconsCommandableGrpcClientV1 : CommandableGrpcClient, IBeaconsClientV1
	{
		public BeaconsCommandableGrpcClientV1()
			: base("beacons_v1")
		{
		}

		public async Task<CenterObjectV1> CalculatePositionAsync(string correlationId, string siteId, string[] udis)
		{
			correlationId ??= IdGenerator.NextLong();

			return await CallCommandAsync<CenterObjectV1>("calculate_position", correlationId, new
			{
				site_id = siteId,
				udis = udis
			});
		}

		public async Task<BeaconV1> CreateBeaconAsync(string correlationId, BeaconV1 beacon)
		{
			correlationId ??= IdGenerator.NextLong();

			return await CallCommandAsync<BeaconV1>("create_beacon", correlationId, new
			{
				beacon = beacon
			});
		}

		public async Task<BeaconV1> UpdateBeaconAsync(string correlationId, BeaconV1 beacon)
		{
			correlationId ??= IdGenerator.NextLong();

			return await CallCommandAsync<BeaconV1>("update_beacon", correlationId, new
			{
				beacon = beacon
			});
		}

		public async Task<BeaconV1> DeleteBeaconByIdAsync(string correlationId, string id)
		{
			correlationId ??= IdGenerator.NextLong();

			return await CallCommandAsync<BeaconV1>("delete_beacon_by_id", correlationId, new
			{
				beacon_id = id
			});
		}

		public async Task<BeaconV1> GetBeaconByIdAsync(string correlationId, string id)
		{
			correlationId ??= IdGenerator.NextLong();

			return await CallCommandAsync<BeaconV1>("get_beacon_by_id", correlationId, new
			{
				beacon_id = id
			});
		}

		public async Task<BeaconV1> GetBeaconByUdiAsync(string correlationId, string udi)
		{
			correlationId ??= IdGenerator.NextLong();

			return await CallCommandAsync<BeaconV1>("get_beacon_by_udi", correlationId, new
			{
				udi = udi
			});
		}

		public async Task<DataPage<BeaconV1>> GetBeaconsAsync(string correlationId, FilterParams filter, PagingParams paging)
		{
			correlationId ??= IdGenerator.NextLong();

			return await CallCommandAsync<DataPage<BeaconV1>>("get_beacons", correlationId, new
			{
				filter,
				paging
			});
		}
	}
}
